### PR TITLE
Feature: Encoders/decoders for token minting and wallet management

### PR DESCRIPTION
**Pull request by josemarinas**
## Description

Add encoders and decoders for:
- Token minting
- Add members
- remove members

Task: [ID]()

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder of the package after the [UPCOMING] title and before the latest version.
- [ ] I have tested my code on the test network.

### CODE REVIEW

File: javascript/modules/client/README.md [hunk 1]
```diff
@@ -1762,6 +1762,101 @@ const configAction = client.encoding.updatePluginSettingsAction(
...
+const contextPlugin: ContextPlugin = ContextPlugin.fromContext(context);
+const client = new ClientErc20(contextPlugin);
+
+const params: IMintTokenParams = {
+  address: "0x1234567890123456789012345678901234567890",
+  amount: BigInt(10),
+};
+
+const tokenAddress = "0x0987654321098765432109876543210987654321";
+const action = client.encoding.mintTokenAction(tokenAddress, params);
```
Comment (brickpop) [inline javascript/modules/client/README.md]: Can we confirm on the existing contract code whether minting only accepts a single entry/address per call?

Comment (brickpop) [inline javascript/modules/client/README.md]:
```suggestion
const pluginAddress = "0x0987654321098765432109876543210987654321";
const action = client.encoding.mintTokenAction(pluginAddress, params);
```

File: javascript/modules/client/README.md [hunk 2]
```diff
@@ -1762,6 +1762,101 @@ const configAction = client.encoding.updatePluginSettingsAction(
...
+console.log(action);
+/*
+{
+  to: "0x1234567890...",
```
Comment (brickpop) [inline javascript/modules/client/README.md]: Shouldn't this address look like `0x098765432109876...` per the example above??

File: javascript/modules/client/README.md [hunk 3]
```diff
@@ -1762,6 +1762,101 @@ const configAction = client.encoding.updatePluginSettingsAction(
...
+import {
+  ClientErc20,
+  Context,
+  ContextPlugin,
+  IMintTokenParams,
+} from "@aragon/sdk-client";
+
+const context: Context = new Context(params);
+const contextPlugin: ContextPlugin = ContextPlugin.fromContext(context);
+const client = new ClientErc20(contextPlugin);
```
Comment (brickpop) [inline javascript/modules/client/README.md]:
```suggestion
const client = new ClientAddressList(contextPlugin);
```

File: javascript/modules/client/README.md [hunk 4]
```diff
@@ -1921,6 +2016,26 @@ console.log(params)
 }
 */
 ```
+### Decode Mint Token Action (ERC-20)
```
Comment (brickpop) [inline javascript/modules/client/README.md]: Why are the 3 new encoding actions together but not the 3 new decoding ones? Adding things in a predictable place makes them easier to find.

File: javascript/modules/client/src/internal/encoding/plugins.ts [hunk 5]
```diff
@@ -137,3 +137,97 @@ function wrapUpdatePluginSettings(result: Result): IPluginSettings {
...
+    amount: BigInt(result[1]),
+  };
+}
+
+// ADD MEMBERS ACTION
+export function encodeAddMembersAction(members: string[]): Uint8Array {
+  const votingInterface = WhitelistVoting__factory.createInterface();
+  // get hex bytes
+  const hexBytes = votingInterface.encodeFunctionData(
+    "addWhitelistedUsers",
```
Comment (brickpop) [inline javascript/modules/client/src/internal/encoding/plugins.ts:181]:
```suggestion
    // TODO: Rename to `addAddresses` as soon as the plugin is updated
    "addWhitelistedUsers",
```

File: javascript/modules/client/src/internal/encoding/plugins.ts [hunk 6]
```diff
@@ -137,3 +137,97 @@ function wrapUpdatePluginSettings(result: Result): IPluginSettings {
...
+  const hexBytes = bytesToHex(data, true);
+  const receivedFunction = votingInterface.getFunction(
+    hexBytes.substring(0, 10) as any,
+  );
+  const expectedfunction = votingInterface.getFunction("addWhitelistedUsers");
+  if (receivedFunction.name !== expectedfunction.name) {
+    throw new UnexpectedActionError();
+  }
+  const result = votingInterface.decodeFunctionData(
+    "addWhitelistedUsers",
```
Comment (brickpop) [inline javascript/modules/client/src/internal/encoding/plugins.ts:202]:
```suggestion
    // TODO: Rename to `addAddresses` as soon as the plugin is updated
    "addWhitelistedUsers",
```

File: javascript/modules/client/src/internal/encoding/plugins.ts [hunk 7]
```diff
@@ -137,3 +137,97 @@ function wrapUpdatePluginSettings(result: Result): IPluginSettings {
...
+    data,
+  );
+  return result[0];
+}
+// REMOVE MEMBERS ACTION
+export function encodeRemoveMembersAction(members: string[]): Uint8Array {
+  const votingInterface = WhitelistVoting__factory.createInterface();
+  // get hex bytes
+  const hexBytes = votingInterface.encodeFunctionData(
+    "removeWhitelistedUsers",
```
Comment (brickpop) [inline javascript/modules/client/src/internal/encoding/plugins.ts:213]:
```suggestion
    // TODO: Rename to `removeAddresses` as soon as the plugin is updated
    "removeWhitelistedUsers",
```

File: javascript/modules/client/src/internal/encoding/plugins.ts [hunk 8]
```diff
@@ -137,3 +137,97 @@ function wrapUpdatePluginSettings(result: Result): IPluginSettings {
...
+export function decodeRemoveMemebersAction(
+  data: Uint8Array,
+): string[] {
+  const votingInterface = WhitelistVoting__factory.createInterface();
+  const hexBytes = bytesToHex(data, true);
+  const receivedFunction = votingInterface.getFunction(
+    hexBytes.substring(0, 10) as any,
+  );
+  const expectedfunction = votingInterface.getFunction(
+    "removeWhitelistedUsers",
```
Comment (brickpop) [inline javascript/modules/client/src/internal/encoding/plugins.ts:228]:
```suggestion
    // TODO: Rename to `removeAddresses` as soon as the plugin is updated
    "removeWhitelistedUsers",
```

File: javascript/modules/client/src/internal/encoding/plugins.ts [hunk 9]
```diff
@@ -137,3 +137,97 @@ function wrapUpdatePluginSettings(result: Result): IPluginSettings {
...
+  );
+  if (receivedFunction.name !== expectedfunction.name) {
+    throw new UnexpectedActionError();
+  }
+  const result = votingInterface.decodeFunctionData(
+    "removeWhitelistedUsers",
```
Comment (brickpop) [inline javascript/modules/client/src/internal/encoding/plugins.ts:235]:
```suggestion
    // TODO: Rename to `removeAddresses` as soon as the plugin is updated
    "removeWhitelistedUsers",
```

File: javascript/modules/client/src/internal/encoding/plugins.ts [hunk 10]
```diff
@@ -137,3 +137,97 @@ function wrapUpdatePluginSettings(result: Result): IPluginSettings {
...
+
+export function decodeAddMemebersAction(
+  data: Uint8Array,
+): string[] {
+  const votingInterface = WhitelistVoting__factory.createInterface();
+  const hexBytes = bytesToHex(data, true);
+  const receivedFunction = votingInterface.getFunction(
+    hexBytes.substring(0, 10) as any,
+  );
+  const expectedfunction = votingInterface.getFunction("addWhitelistedUsers");
```
Comment (brickpop) [inline javascript/modules/client/src/internal/encoding/plugins.ts:196]:
```suggestion
  // TODO: Rename to `addAddresses` as soon as the plugin is updated
  const expectedfunction = votingInterface.getFunction("addWhitelistedUsers");
```

File: javascript/modules/client/test/integration/client-addressList.test.ts [hunk 11]
```diff
@@ -422,6 +425,116 @@ describe("Client", () => {
...
+        "0x1357924680135792468013579246801357924680",
+        "0x2468013579246801357924680135792468013579",
+        "0x0987654321098765432109876543210987654321",
+      ];
+
+      const pluginAddress = "0x1234567890123456789012345678901234567890";
+      const action = client.encoding.addMembersAction(pluginAddress, members);
+
+      expect(typeof action).toBe("object");
+      expect(action.data).toBeInstanceOf(Uint8Array);
```
Comment (brickpop) [inline javascript/modules/client/test/integration/client-addressList.test.ts:480]: Maybe it's excessive but I would suggest to convert `action.data` into a hex string, and compare it to the hardcoded expected value

File: javascript/modules/client/test/integration/client-addressList.test.ts [hunk 12]
```diff
@@ -422,6 +425,116 @@ describe("Client", () => {
...
+      ];
+
+      const pluginAddress = "0x1234567890123456789012345678901234567890";
+      const action = client.encoding.removeMembersAction(
+        pluginAddress,
+        members,
+      );
+
+      expect(typeof action).toBe("object");
+      expect(action.data).toBeInstanceOf(Uint8Array);
```
Comment (brickpop) [inline javascript/modules/client/test/integration/client-addressList.test.ts:539]: Same as above

File: javascript/modules/client/test/integration/client-erc20.test.ts [hunk 13]
```diff
@@ -409,10 +412,55 @@ describe("Client", () => {
...
+
+      const tokenAddress = "0xinvalid_address";
+      expect(() =>
+        client.encoding.mintTokenAction(
+          tokenAddress,
+          params,
+        )
+      ).toThrow(new InvalidAddressError());
+    });
+    it("Should encode a mint action", async () => {
```
Comment (brickpop) [inline javascript/modules/client/test/integration/client-erc20.test.ts]:
```suggestion
    it("Should encode an ERC20 token mint action", async () => {
```

File: javascript/modules/client/test/integration/client-erc20.test.ts [hunk 14]
```diff
@@ -409,10 +412,55 @@ describe("Client", () => {
...
+      const client = new ClientErc20(context);
+      const params: IMintTokenParams = {
+        address: "0x1234567890123456789012345678901234567890",
+        amount: BigInt(10),
+      };
+
+      const tokenAddress = "0x0987654321098765432109876543210987654321";
+      const action = client.encoding.mintTokenAction(tokenAddress, params);
+      expect(typeof action).toBe("object");
+      expect(action.data).toBeInstanceOf(Uint8Array);
```
Comment (brickpop) [inline javascript/modules/client/test/integration/client-erc20.test.ts:462]: Same as above

File: javascript/modules/client/README.md (hunk 1 above)
Comment (josemarinas) [reply javascript/modules/client/README.md]: https://github.com/aragon/core/blob/73f84b6314e3caa0a475b8172e7ec90fe8df91e3/packages/contracts/contracts/tokens/IERC20MintableUpgradeable.sol

Seems like it, unless this is not the correct function

Comment (josemarinas) [reply javascript/modules/client/README.md]: Why?, correct me if I'm wrong but here you don't need the plugin address, only the token address

Comment (brickpop) [reply javascript/modules/client/README.md]: Well, we are encoding an action that the DAO will execute. Technically speaking, the DAO should call `mint()` on the token minter (which could be the token contract itself). I personally prefer to design things such in a way that only plugins talk to the DAO and vice versa. But in any case, the entity minting tokens is the minter... let's consider `minterAddress`instead

File: javascript/modules/client/test/integration/client-addressList.test.ts [hunk 15]
```diff
@@ -475,8 +476,11 @@ describe("Client", () => {
       const pluginAddress = "0x1234567890123456789012345678901234567890";
       const action = client.encoding.addMembersAction(pluginAddress, members);
 
+      const functionFragment = getFunctionFragment(action.data)
+
       expect(typeof action).toBe("object");
       expect(action.data).toBeInstanceOf(Uint8Array);
+      expect(functionFragment.name).toBe("addWhitelistedUsers");
```
Comment (brickpop) [inline javascript/modules/client/test/integration/client-addressList.test.ts]: Isn't it easier to simply do...?
```suggestion
      expect(bytesToHex(action.data)).toBe("0x.....__hardcoded_here__");
```

Comment (brickpop) [reply javascript/modules/client/test/integration/client-addressList.test.ts]: The goal is preventing accidental changes. We don't need to test the decoding here

File: javascript/modules/client/test/integration/client-addressList.test.ts [hunk 16]
```diff
@@ -475,8 +476,11 @@ describe("Client", () => {
       const pluginAddress = "0x1234567890123456789012345678901234567890";
       const action = client.encoding.addMembersAction(pluginAddress, members);
 
+      const functionFragment = getFunctionFragment(action.data)
```
Comment (brickpop) [inline javascript/modules/client/test/integration/client-addressList.test.ts]: See below

File: javascript/modules/client/test/integration/client-addressList.test.ts [hunk 17]
```diff
@@ -530,9 +534,11 @@ describe("Client", () => {
         pluginAddress,
         members,
       );
+      const functionFragment = getFunctionFragment(action.data)
```
Comment (brickpop) [inline javascript/modules/client/test/integration/client-addressList.test.ts]: Same as above

File: javascript/modules/client/test/integration/client-addressList.test.ts [hunk 18]
```diff
@@ -530,9 +534,11 @@ describe("Client", () => {
...
 
       expect(typeof action).toBe("object");
       expect(action.data).toBeInstanceOf(Uint8Array);
+      expect(functionFragment.name).toBe("removeWhitelistedUsers");
```
Comment (brickpop) [inline javascript/modules/client/test/integration/client-addressList.test.ts]: Same as above

File: javascript/modules/client/test/integration/client-erc20.test.ts [hunk 19]
```diff
@@ -495,7 +496,10 @@ describe("Client", () => {
       const tokenAddress = "0x0987654321098765432109876543210987654321";
       const action = client.encoding.mintTokenAction(tokenAddress, params);
       const decodedParams = client.decoding.mintTokenAction(action.data);
+      const functionFragment = getFunctionFragment(action.data);
+
       expect(decodedParams.address).toBe(params.address);
+      expect(functionFragment.name).toBe("mint");
```
Comment (brickpop) [inline javascript/modules/client/test/integration/client-erc20.test.ts]: Same as above

**Review by brickpop — APPROVED** [review]

**Review by brickpop — APPROVED** [review]

**Pull request merged**